### PR TITLE
support chinese file name

### LINKED ISSUE #34: Support for Chinese characters
https://github.com/junrar/junrar/blob/ae7bd9bb0590a530ff8b2057e1a0c403c004085c/src/main/java/com/github/junrar/Junrar.java#L159

使用`fileHeader.getFileNameString()`获取FileName中文会乱码，而使用
`fileHeader.getFileNameW()`
就不乱码

### PR DESCRIPTION
**Pull request by sunny-shu**
fix #34

### CODE REVIEW

File: src/main/java/com/github/junrar/rarfile/FileHeader.java [hunk 1]
```diff
@@ -159,6 +159,10 @@ public FileHeader(BlockHeader bh, byte[] fileHeader) {
                 fileName = new String(fileNameBytes);
                 fileNameW = "";
             }
+            // support Chinese characters
+            if (fileNameW.length() > 0) {
```
Comment (gotson) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:163]: what is `fileNameW` exactly?

**Comment by rednoah** [comment]
Note that using `new String(byte[])` to decode external strings is also potentially dangerous, since it may very well work on some platforms, but not others, depending whatever the platform's default charset happens to be:
https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#String-byte:A-

AFAIK, `fileName` is 1-byte-per-1-character fixed length ASCII, while `fileNameW` typically refers to 2-byte-per-1-character (but possibly variable-length) UTF-16 encoding.

File: src/main/java/com/github/junrar/rarfile/FileHeader.java (hunk 1 above)
Comment (sunny-shu) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:163]: public boolean isUnicode() {
        return (flags & LHD_UNICODE) != 0;
    }

if LHD_UNICODE is set,the file name field contain two part.first part is the bytes of file name encoded by local charset. the second part is the bytes of file name encoded  by unicode. Two parts are separated by 0。
so i think fileNameW  represents the file name which is decoded from second part   in unicode。
String use utf-8 to decode bytes by default,so fileName garbled when the file name is Chinese

**Comment by sunny-shu** [comment]
@rednoah 
what is the potential danger?
you means security question or just Incorrect behavior？

**Comment by gotson** [comment]
Does the issue happen with files created with winrar or with rar from the command-line?

**Comment by sunny-shu** [comment]
@gotson 
yes, i use winrar to pack a file with chinese name.,and test  the class Archive with it. 
then the fileName is  garbled。

File: src/test/java/com/github/junrar/ArchiveTest.java [hunk 2]
```diff
@@ -62,6 +62,15 @@ public void testTikaDocs() throws Exception {
         }
     }
 
+    @Test
+    public void chineseFileName() throws Exception {
+        File f = new File(getClass().getResource("password/rar4-only-file-content-encrypted.rar").getPath());
+        try (Archive archive = new Archive(f, "test")) {
+            FileHeader fileHeader = archive.nextFileHeader();
+            assertThat(fileHeader.getFileNameString().equals("新建文本文档.txt"));
```
Comment (gotson) [inline src/test/java/com/github/junrar/ArchiveTest.java:70]: This is incorrect and always return true, it should be:
`assertThat(fileHeader.getFileNameString()).isEqualTo("新建文本文档.txt");`

**Comment by gotson** [comment]
I just checked the code, and to me it already works. The thing is that the caller should use `getFileNameW()` instead of `getFileNameString()`.

`LocalFolderExtractor` already handles it this way:
```java
if (fh.isUnicode()) {
  fileName = fh.getFileNameW();
} else {
  fileName = fh.getFileNameString();
}
```

The accessors are not really well name, and probably follow the `unrar` source code. Instead of changing the code like you did, we could add a new accessor `getFileName()` which would return `filenameW` if its length is > 0, and `filename` else.

We could also rename the 2 existing accessors for further clarification:
- `getFileNameString()` => `getFileNameAscii()`
- `getFileNameW()` => `getFileNameUnicode()`

What do you think?

**Comment by gotson** [comment]
I re-read the original issue, and it actually describes a working behavior:

> Use fileHeader.getFileNameString() to get FileName Chinese will be garbled, and use fileHeader.getFileNameW() Not garbled (google translated)

I do agree it's not intuitive though, hence my proposal above.

**Comment by theobisproject** [comment]
I think it would be better to deprecate the exisiting `getFileName*` methods and implement one like this as replacement
```java
public String getFileName() {
    return isUnicode() ? getFileNameW() : getFileNameString();
}
```
Then the user does not need to care if the file name is in Unicode or not. Also I no longer have to do it in various places in my application 😉

**Comment by gotson** [comment]
Would there be cases however where one would want to retrieve the ASCII filename while there is a Unicode filename?

**Comment by theobisproject** [comment]
Might be possible that someone only wants to use the ASCII filename but I'm currently unable to think of one. Therefore I proposed only to deprecate the methods and wait some time before removing them. So anyone who needs those methods has time to stand up.

**Comment by gotson** [comment]
Sounds good, I agree with that. We can alway re-add the ASCII one later on with a proper name if needed.

**Comment by sunny-shu** [comment]
@gotson @theobisproject 
public String getFileName() {
    return isUnicode() ? getFileNameW() : getFileNameString();
}
i think this will be nice.
As a user，they don't know how winrar encode the file name, and they don't care about it.

**Comment by gotson** [comment]
i'll do the change this morning, i will add the deprecation, and update our own code to use the new method.

**Pull request closed without merging**